Fix VirtualizedList with maintainVisibleContentPosition

janicduplessis opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary

`maintainVisibleContentPosition` is broken when using virtualization and the new content pushes visible content outside its "window". This can be reproduced in the example from this diff. When using a large page size it will always push visible content outside of the list "window" which will cause currently visible views to be unmounted so the implementation of `maintainVisibleContentPosition` can't adjust the content inset since the visible views no longer exist.

The first illustration shows the working case, when the new content doesn't push visible content outside the window. The red box represents the window, all views outside the box are not mounted, which means the native implementation of `maintainVisibleContentPosition`  has no way to know it exists. In that case the first visible view is #2, after new content is added #2 is still inside the window so there's not problem adjusting content offset to maintain position. As you can see Step 1 and 3 result in the same position for all initial views.

The second illustation shows the broken case, when new content is added and pushes the first visible view outside the window. As you can see in step 2 the view #2 is no longer rendered so there's no way to maintain its position.

#### Illustration 1

![image](https://user-images.githubusercontent.com/2677334/163263472-eaf7342a-9b94-4c49-9a34-17bf8ef4ffb9.png)

#### Illustration 2

![image](https://user-images.githubusercontent.com/2677334/163263528-a8172341-137e-417e-a0c7-929d1e4e6791.png)

To fix `maintainVisibleContentPosition` when using `VirtualizedList` we need to make sure the visible items stay rendered when new items are added at the start of the list.

In order to do that we need to do the following:

- Detect new items that will cause content to be adjusted
- Add cells to render mask so that previously visible cells stay rendered
- Ignore certain updates while scroll metrics are invalid

### Detect new items that will cause content to be adjusted

The goal here is to know that scroll position will be updated natively by the `maintainVisibleContentPosition` implementation. The problem is that the native code uses layout heuristics which are not easily available to JS to do so. In order to approximate the native heuristic we can assume that if new items are added at the start of the list, it will cause `maintainVisibleContentPosition` to be triggered. This simplifies JS logic a lot as we don't have to track visible items. In the worst case if for some reason our JS heuristic is wrong, it will cause extra cells to be rendered until the next scroll event, or content position will not be maintained (what happens all the time currently). I think this is a good compromise between complexity and accuracy.

We need to find how many items have been added before the first one. To do that we save the key of the first item in state `firstItemKey`. When data changes we can find the index of `firstItemKey` in the new data and that will be the amount we need to adjust the window state by.

Note that this means that keys need to be stable, and using index won't work.

### Add cells to render mask so that previously visible cells stay rendered

Once we have the adjusted number we can save this in a new state value `maintainVisibleContentPositionAdjustment` and add the adjusted cells to the render mask.

This state is then cleared when we receive updated scroll metrics, once the native implementation is done adding the new items and adjusting the content offset.

This value is also only set when `maintainVisibleContentPosition` is set so this makes sure this maintains the currently behavior when that prop is not set.

### Ignore certain updates while scroll metrics are invalid

While the `maintainVisibleContentPositionAdjustment` state is set we know that the current scroll metrics are invalid since they will be updated in the native `ScrollView` implementation. In that case we want to prevent certain code from running.

One example is `onStartReached` that will be called incorrectly while we are waiting for updated scroll metrics.

## Changelog

[General] [Fixed] - Fix VirtualizedList with maintainVisibleContentPosition

## Test Plan

Added bidirectional paging to RN tester FlatList example. Note that for this to work RN tester need to be run using old architecture on iOS, to use new architecture it requires https://github.com/facebook/react-native/pull/35319

Using debug mode we can see that virtualization is still working properly, and content position is being maintained.

https://user-images.githubusercontent.com/2677334/163294404-e2eeae5b-e079-4dba-8664-ad280c171ae6.mov

janicduplessis commented:
Working on adding some tests

NickGerleman commented:
Very excited by this.

cc @rozele as well who was looking at scroll anchoring.

This should solve a huge class of issues with relayout pushing items around during realization. But it also actually would allow us to implement APIs like scrollToEnd() in a way which avoids flutter as well (since as of 0.71+ we can realize the end cell ahead of time and use as an anchor if we scroll down before other content is rendered.).

Going to try to dedicate some time to read through all of this. Have a feeling this will be one of those changes that requires a cup of coffee and some quiet time to grep through 😛.

mikfoo commented:
I've been testing this change for a while and for the most part it works great. However, sometimes when adding a lot of items at once it still seems to lose the anchor. I've recorded a short video to demonstrate the issue.

In the video it works great the first three times I reset the app and add items to the list, but then fails and loses anchor after more resets when adding items.

Each time I press the red square, `50` items are prepended to the list, each with a height of `200`. The purple badge in the right corner shows the item count. After each reset it starts at 4 items. When adding fewer items at a time (eg. 5) it never seems to lose its anchor.

https://user-images.githubusercontent.com/144865/220893160-6013f8ca-9650-43de-b23e-a974f0d36b2f.mp4

janicduplessis commented:
@mikfoo Could you provide me with the code you used to reproduce this problem?

mikfoo commented:
> @mikfoo Could you provide me with the code you used to reproduce this problem?

If you create a new RN project and apply the patch manually to `VirtualizedList.js` in `node_modules` and prepend data to a simple FlatList the issue occurs, at least for me. Here's a quick example you can use: https://gist.github.com/mikfoo/289e8c54e04f0a34f090ce10e3603d90

With the code in the gist I can reliably reproduce the issue over and over. It usually happens on first "Add" and also when you have scrolled to the top of the list (when `contentOffset.y <= 0`).

Another small thing I noticed (not sure if it's directly related to this PR) is that when you add new items too fast to the list (eg. pressing the "Add" button a couple of times in fast succession) then it throws an error:

<img width="444" alt="image" src="https://user-images.githubusercontent.com/144865/221397030-0d295a3d-c27d-4a08-a283-325252311dcd.png">

rozele commented:
I *think* this will also work for the inverted case on mobile platforms, since that just uses transform tricks to make the start of the list render at the bottom, but it might be worth validating. For platforms like Web and Windows, we actually need to change the way inversion works because the transform tricks don't play nice with keyboard nav, mouse wheel events, etc. Anecdotally, on Windows, we have a prototype for this (https://github.com/microsoft/react-native-windows/pull/8440).

It feels a bit like the hacks to get things to work in VirtualizedList are piling up 😅. I wonder how many of these could be resolved with a native virtualized list component (render mask for retaining focused islands, maintainVisibleContentPosition virtualization, native inversion for relevant platforms)...

rozele commented:
I also wonder if you could build on @NickGerleman's render mask functionality to implement this. E.g., add a new event when a scroll anchor is selected and force an "island" of materialized cells around the selected anchor. Not sure it would be feasible, given the virtualization logic runs from onLayout and onScroll, and you don't select an anchor for maintainVisibleContentPosition until ScrollView layout (which would be too late to prevent the bug).

I've also long wondered if there's a better alternative to maintainVisibleContentPosition. Here are some ideas I've had in the past:
1. Switch the default behavior of React Native to anchor-by-default as on Web (this wouldn't fix this particular bug, but still perhaps a good idea).
2. Add an `overflowAnchor` prop to View to enable / disable anchoring for specific children of ScrollView.
3. (Fabric-only) Add a synchronous event to ScrollView, like `onAnchorRequested` to allow app code to specify the anchor just before layout is committed on the UI thread (there's a corollary to this on Windows with the ScrollViewer::AnchorRequested event).

If we switched to "anchor-by-default" behavior and assumed that an anchor was selected on every commit, we could be certain an event would fire with the selected anchor in VirtualizedList 🤷‍♂️.

This is all just food for thought. I have no opposition to fixing bugs for code that should already work 😅

mikfoo commented:
@janicduplessis I played around with it again last night, and noticed that if you add items while it's batch rendering then it always loses focus. If you wait adding items to the list until it's completely done rendering, then the focus is kept (except on the initial render, it always loses focus on that).

janicduplessis commented:
@mikfoo Thanks for testing this! I was able to reproduce the issue with the code you provided and it should now be fixed with 3379aec2f12c121875b3d949340a6caa1bbca1f3. In the example I used to test I added items during scrolling with `onStart/EndReached` which would always trigger high priority updates which don't have that problem. To fix it I can cancel pending batches (which are now invalid because of the updated maintainVisibleContentPositionAdjustment) when the adjustment changes.

I can also reproduce the issue where the first time items are added the adjustment doesn't work. Will have a look at this next.

janicduplessis commented:
@rozele Thanks for having a look. I agree with you that a better long term solution would be great to explore, but in the meantime this is an acceptable fix.

I am pretty sure inverted works, but will double check.

> I also wonder if you could build on @NickGerleman's render mask functionality to implement this. E.g., add a new event when a scroll anchor is selected and force an "island" of materialized cells around the selected anchor.

Unless I'm wrong about what you mean here, this is already what I do [here](https://github.com/facebook/react-native/pull/35993/files#diff-b450c5c4ec05ca5221b18a829f764b8397eed35a93f388893debb184591b0207R546). Basically what I do is detect that the scrollview will be anchored and create an island around previously visible cells.

I think the idea of having anchor events would be great in this case, since currently I have to approximate that a scroll adjustment will happen, since that logic is implemented in native code using layout info not available to JS.

mikfoo commented:
@janicduplessis Awesome! I just tested the changes and they seem to work great! There's a new small issue where when I load a large list, it sometimes doesn't start at the top of the list on the initial load (it's a few items off), but I haven't been able to reliably reproduce it yet.

NickGerleman commented:
I was able to take a look through the code, and description. I have some implementation comments, but before I leave those,  I wanted to check to see if I understand the overall problem, and the viability of an alternative method of doing this.

From the Flow, here, it seems like:
1. User adds data before the viewport
2. VirtualizedList responds to the new data, by rendering an index-based window around the new data
3. `maintainVisibleContentPosition` doesn't save us and adjust scroll position because the old item isn't still rendered

What I am wondering, could we attack #2 by changing how VirtualizedList handles `data` changes.

I.e. fix existing logic in `getDerivedState()` to operate on `key` instead of index when determining the next render range on prop change.

NickGerleman commented:
A comment pending on this implementation, that I think would effect that one as well, is that we usually want to avoid the full scan of `data` like this change is currently doing. The model is meant to allow `data` to be backed by something expensive/lazy, so doing a full search throughout all data defeats that.

Though I am not sure the current API provides a way to diff before/after efficiently, without searching for keys. But we could also maybe optimize that search (e.g. look around the last index first).

janicduplessis commented:
@NickGerleman Thanks for having a look, if I understand your point correctly you suggest adjusting the `cellsAroundViewport` [here](https://github.com/facebook/react-native/blob/f673c7d13668094b22eba33e41d61dc4d0c8a0cd/packages/virtualized-lists/Lists/VirtualizedList.js#L782) instead of rendering an extra window based on `maintainVisibleContentPositionAdjustment`? I had something similar in my original implementation in the old VirtualizedList that didn't support multiple render windows, but I figured it would be better to leverage multiple windows when doing the adjustment so in case the adjustment does not happen in native (could be possible since we approximate that the adjustment will happen by checking item changes while native uses visible items layout). If the adjustment does not happen by having both old and new window rendered it would avoid rendering blank until the user scrolls and right items are rendered again. Since the larger range will only be rendered for a very short amount of time I figured that this should be ok.

> Though I am not sure the current API provides a way to diff before/after efficiently, without searching for keys. But we could also maybe optimize that search (e.g. look around the last index first).

Yea the best I can think of is to use the change in data size and start the lookup from there. So if items are added at the start the old first item will be at index `newItemCount - oldItemCount - 1` so we would get it at first iteration. However if items are added at the end of the list then we would need to scan more. We could try checkout index 0 and `newItemCount - oldItemCount - 1` first and fallback to full scan. Note that this case only happens when mvcp is set so it doesn't affect performance of previous VirtualizedList usage.

NickGerleman commented:
> I had something similar in my original implementation in the old VirtualizedList that didn't support multiple render windows, but I figured it would be better to leverage multiple windows when doing the adjustment so in case the adjustment does not happen in native (could be possible since we approximate that the adjustment will happen by checking item changes while native uses visible items layout). If the adjustment does not happen by having both old and new window rendered it would avoid rendering blank until the user scrolls and right items are rendered again. 

My recommendation is that we shift the old window, to a single new location, instead of rendering two windows. Then we explicitly rely on `maintainVisibleContentPosition` to keep the scroll position aligned.

On data change, VirtualizedList's strategy of rendering same index, instead of same key, means potentially very expensive renders that are not where we want. Say, our window was items 20-40 of a 100 item list, and we erase items 0-20, we end up rendering 20 new items. With FlatList defaults, this means we end up rendering 21 viewports worth of items, as a waste.

Now, with the current new logic, we add another 21 potentially new viewports. For context, the scenario around keeping the last focus cell realized will only keep a single viewport above/below the cell we need to keep alive, which is sufficient to avoid blanking without expensive renders. We still need native `maintainVisibleContentPosition` as well for alignment, since the new window may have elements of different measure than the spacer. 

If we instead realign `cellAroundViewport` based on previous keys, we can behave much more cheaply, and the logic is all confined to the invalidation step, instead of what we have right now which hooks deeply into different facets of the list.

> Yea the best I can think of is to use the change in data size and start the lookup from there. So if items are added at the start the old first item will be at index newItemCount - oldItemCount - 1 so we would get it at first iteration. However if items are added at the end of the list then we would need to scan more. We could try to check index 0 and newItemCount - oldItemCount - 1 first and fallback to full scan.

For the reasons above, I think we may want to do this realignment in more general cases. Thinking more about this, it should be possible to do efficiently, and with a bounds. Just slower compared to us having information ahead of time.

1. Add all previous `cellsAroundViewport` cellKeys to a set, then iterate through the new window to create an mapping from old index to new index.
  a. This bounds to `windowSize`, where we already may do a full scan of that length (though usually on hot data)
  b. We can do things like short circuit if we find the previous anchor cell we were using
2. Realign new `cellsAroundViewport` based on:
  a. An explicit `maintainVisibleContentPosition` anchor if mappable
  b. Some other deterministically chosen anchor from the set

That kinda veers on having `maintainVisibleContentPosition` enabled on invalidation, since our logic is explicitly allowing `cellsAroundViewport` to drift in layout space to a greater degree.

NickGerleman commented:
Though also FYI @rozele @acoates-ms that would add a dependency on maintainVisibleContentPosition, which I think will soon only be a concern for Windows.

janicduplessis commented:
@NickGerleman Hopefully I got your feedback right, as I'm not familiar with all of VirtualizedList code.

I changed it so that instead of creating a 2nd window with the mvcp adjustment I instead update the `cellsAroundViewport`. 

> For the reasons above, I think we may want to do this realignment in more general cases. Thinking more about this, it should be possible to do efficiently, and with a bounds in common cases. Just slower compared to us having information ahead of time.

I'm not sure I understand here what you mean by doing a realignment in more general cases, also the part about creating a mapping from old to new index `2.a An explicit maintainVisibleContentPosition anchor if mappable`. Using the current technique to detect changes that would trigger native `maintainVisibleContentPosition`, we only rely on the first item in the list, which might not be in the currently rendered window, which is why I'm not sure I understand why a mapping using the window would help in that case.

I added a hint to `_findItemIndexWithKey` so we can find the item index in O(1) when adding items at the start of the list instead of the number of new items iterations. Unless the first item is removed from the list or content added at the start and another position at the same time (so our length hint is off) we don't need to do any scanning of data.

NickGerleman commented:
> I'm not sure I understand here what you mean by doing a realignment in more general cases

The idea is that we adjust `cellsAroundViewport` on data change using a key-based anchor, instead of the index based anchor, even when we are not explicitly using `maintainVisibleContentPosition`. This avoids some special casing, but should also fix pre-existing jank around data changes. It may be more involved than we want though (I will take a look through this PR).

NickGerleman commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -712,14 +747,43 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
     return {
       cellsAroundViewport: constrainedCells,
       renderMask: VirtualizedList._createRenderMask(newProps, constrainedCells),
+      firstItemKey: newFirstItemKey,
+      maintainVisibleContentPositionAdjustment,
```
It seems like:
1. We never use the adjustment value outside of here, apart from as a check as to whether we have an updated scroll position yet
2. We do not use `firstItemKey` outside of this function.

So could we instead have simpler state like: `pendingScrollPositionUpdate: boolean`? Or `pendingScrollUpdateCount: number` (see below).

janicduplessis replied:
We do need `firstItemKey` as state to check if the value changed.

NickGerleman commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -1643,6 +1718,11 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
     };
+    if (this.state.maintainVisibleContentPositionAdjustment != null) {
+      this.setState({
+        maintainVisibleContentPositionAdjustment: null,
```
This may not be correct in the following sequence:
1. We change data once
2. We change data again
3. We get a scroll update from #&#8203;1

Consider a counter instead like `pendingScrollUpdateCount: number`.

janicduplessis replied:
Good catch on this, I was able to reproduce this case by adding pages multiple times with a small delay

```js
const onPress = () => {
  addPages();
  setTimeout(() => {
    addPages();
  }, 1);
}

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -502,6 +509,34 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
+  }
+
+  static _getItemKey(props: Props, index: number): ?string {
+    const item = props.data ? props.getItem(props.data, index) : null;
```
This would cause `renders null list` test to fail since we use it to set initial state. I could move the null check to problematic call site too.

Not sure exactly what you mean by `There is code out there with nullish data but getItem hardcoding values.`

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -1121,6 +1184,15 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
     if (hiPriInProgress) {
       this._hiPriInProgress = false;
     }
+
+    // Make sure to cancel any pending updates if maintainVisibleContentPositionAdjustment
+    // changed since they are now invalid.
+    if (
+      prevState.maintainVisibleContentPositionAdjustment !==
+      this.state.maintainVisibleContentPositionAdjustment
+    ) {
+      this._updateCellsToRenderBatcher.dispose({abort: true});
+    }
```
I will double check if this is still needed now that we adjust cellsAroundViewport.

janicduplessis replied:
Will also check the code that you mentioned for handling stale scroll position see if it is possible to leverage it instead of skipping calls.

janicduplessis replied:
This code seems to be still needed, or at least some way to skip updates while we expect a scroll position adjustment since it uses `_scrollMetrics` which will be out of sync with native cause cause a wrong window to be rendered.

janicduplessis replied:
I think folding into `_adjustCellsAroundViewport` will be more complex since there are some other functions like for example `_updateViewableItems` that should not be called with invalid scroll metrics, and it is also called from setState update which I think makes it impossible to use this.state so we'd need to pass `pendingScrollUpdateCount` but at this point might as well just skip calling the function.

One option would be to move the check to `_updateCellsToRender` which would avoid having to check in `_scheduleCellsToRenderUpdate` and cancel pending batches. The check in `_maybeCallOnEdgeReached` is also needed since it is called from many places like `onLayout` which will happen before we get adjusted scroll position.

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -1458,6 +1530,9 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
       onEndReachedThreshold,
       initialScrollIndex,
     } = this.props;
+    if (this.state.maintainVisibleContentPositionAdjustment != null) {
+      return;
+    }
```
This is to avoid triggering incorrect onStartReachedEvents since scroll position is currently invalid. I will double check, but if I remember correctly without this check right after adding new data at the start of the list `onStartReached` would be called, before we get an adjusted scroll position from native, which shouldn't trigger the callback.

In general whenever `maintainVisibleContentPositionAdjustment` is not null, any function which uses scroll position should not be called. This seems to be safe functions that uses scroll position are usually called again whenever we receive the updated scroll position and reset `maintainVisibleContentPositionAdjustment`.

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -712,14 +747,43 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
+    let maintainVisibleContentPositionAdjustment =
+      prevState.maintainVisibleContentPositionAdjustment;
+    const prevFirstItemKey = prevState.firstItemKey;
+    const newFirstItemKey = VirtualizedList._getItemKey(newProps, 0);
```
`minIndexForVisible` is based on ScrollView index of the item, which might be different from the VirtualizedList item index (for example if you have HeaderComponent) so I don't think we should use that value directly. According to the documentation of `minIndexForVisible` it could be used to avoid maintaining scroll position for loading spinners, but I think those will usually be implemented with Header/Footer components of VirtualizedList.

Because of that I think always using the first item is fine in this case.

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -712,14 +747,43 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
+      newFirstItemKey != null
+    ) {
+      // Fast path if items were added at the start of the list.
+      const hint = itemCount - prevState.renderMask.numCells();
```
We won't do the lookup when items are added at the end of the list since the first item did not change (see https://github.com/facebook/react-native/pull/35993/files#diff-b450c5c4ec05ca5221b18a829f764b8397eed35a93f388893debb184591b0207R763). Maybe I should use if .. else block instead of ternary to make this logic more obvious?

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -502,6 +509,34 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
+  static _getItemKey(props: Props, index: number): ?string {
+    const item = props.data ? props.getItem(props.data, index) : null;
+    if (item == null) {
+      return null;
```
The difference between `_getItemKey` and `_keyExtractor` is that the former will take care of calling props.getItem while the latter doesn't and requires passing the item object (see https://github.com/facebook/react-native/pull/35993/files#diff-b450c5c4ec05ca5221b18a829f764b8397eed35a93f388893debb184591b0207R533). I had a couple places where I needed to find an item key by index so this simplified the code, maybe it could be renamed to make it more explicit like `_getItemKeyForIndex` or `_keyExtractorForIndex`?

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -73,6 +73,11 @@ type ViewabilityHelperCallbackTuple = {
 type State = {
```
Yea `getDerivedStateFromProps` makes this more awkward than it should be :/

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -712,14 +747,43 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
 
+    if (maintainVisibleContentPositionAdjustment != null) {
+      constrainedCells.first += maintainVisibleContentPositionAdjustment;
+      constrainedCells.last += maintainVisibleContentPositionAdjustment;
```
I moved the adjustment before the constraining, which should address point 1. Not sure about how this affects point 2 / 3 as we are only sliding the window here and not making it bigger. We are also sliding it back to where it previously was before items were changed so not sure if that makes a difference about amount of rendering.

janicduplessis commented:
@NickGerleman Thanks for having another look, I made a first round of improvements based on your feedback and tried to address all points. LMK if I missed anything.

NickGerleman reviewed (commented):
Thanks for the work you've been putting in here, and the time to help me understand the code. I think we are getting close here.

Apart from some nits, the main piece of feedback is around `minIndexForVisible`. Both:
1. Translating the index to the underlying ScrollView when passed (like we do for `stickyHeaderIndices`).
2. Make the logic checking for detecting when native will shift position, and for choosing the window, aware of the anchor. The logic assuming it is always at zero may create subtlety incorrect results otherwise.

I'm going to be off-the-grid for a few weeks, and would in the meantime recommend staging the the granular elements of this change. This should make it easier to start approving/shipping bits when they are ready, and means it is more likely someone other than me might know what's going on while I'm away.

E.g. One atomic change is that we are wanting to alter the logic for how we handle a known not-yet-correct scroll position. That could be submitted as a PR which adds the changes to no-op at the greater granularity, then updating the existing `initialScrollIndex` case in `adjustCellsToRender`.

Another would be adding the helper over `_keyExtractor` then changing existing usages to use the helper instead.

janicduplessis commented:
@NickGerleman I will also be away for the next few weeks so I will work on addressing your last round of feedback when I am back.

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -879,6 +969,10 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
+            // We expect that header component will be a single native view so make it
+            // not collapsable to avoid this view being flattened and make this assumption
+            // no longer true.
+            collapsable={false}
```
View flattening of header component can causes issues here since we assume header view = 1 native view in minIndexForVisible. I think it is ok to always disable flattening here, but can also do it only if maintainVisibleContentPosition is true.

janicduplessis replied:
This is also why the example had 2 as minIndexForVisible although it would be expected to work with only 1

NickGerleman replied:
Not needed for now, but is the idea here that we need the native component alive around for `maintainVisibleContentPosition` to work?

Is it specifically that the anchor index needs to be visible? Wondering if this needs to live in ScrollView for that case when someone sets `minIndexForVisible`.

janicduplessis replied:
It is more that minIndexForVisible index is based on native view index and not react native view index. So if we have a header view that gets flattened with 2 child view then our header now adds 2 child native views, so our adjustment of 1 view for header is now wrong.

We can’t really make this adjustment in ScrollView easily since collapsable prop needs to be set on ScrollView child views. It would def be possible but I don’t think it is needed.

NickGerleman replied:
> So if we have a header view that gets flattened with 2 child view then our header now adds 2 child native views, so our adjustment of 1 view for header is now wrong.

Ah, thanks for the clarification!

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -616,22 +670,9 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
-      // Thus, we want to recalculate the windowed render limits if any of the following hold:
-      // - initialScrollIndex is undefined or is 0
-      // - initialScrollIndex > 0 AND scrolling is complete
-      // - initialScrollIndex > 0 AND the end of the list is visible (this handles the case
```
Do you know why we need to recalculate the render limits with initialScrollIndex in the case where distanceFromEnd == 0?

With this change I no longer cover this case so I wonder if it is needed.

NickGerleman replied:
I have wondered the same. It seems like its not really doing anything, so I think we can remove if everything still seems like its working.

janicduplessis commented on packages/virtualized-lists/Lists/VirtualizedList.js:
```diff
@@ -1510,14 +1614,8 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
-      // and will trigger an unexpected onStartReached. To avoid this we can use
-      // timestamp to differentiate between the initial scroll metrics and when we actually
-      // received the first scroll event.
-      if (!initialScrollIndex || this._scrollMetrics.timestamp !== 0) {
```
This is no longer needed with pendingScrollUpdateCount check.

janicduplessis commented:
@NickGerleman I think I was able to address your comments. I was able to use pendingScrollUpdateCount to replace checks we had with `initialScrollIndex` by setting the initial value of `pendingScrollUpdateCount` based on `initialScrollIndex` since we also want to skip updating the window until we receive valid scroll metrics.

This did cause some minor changes in snapshots because the value we check is in react state and not an instance variable. I think these snapshots are actually more correct than the previous. 

The previous calls looks like:

- simulateLayout which triggers a call to `_updateCellsToRender`
- simulateScroll which triggers a call to `_updateCellsToRender` and set `this._scrollMetrics`
- setState callback is executed for the first `_updateCellsToRender` -> cells are rendered
- setState callback is executed for the second `_updateCellsToRender` -> cells are rendered again

It now looks like:

- simulateLayout which triggers a call to `_updateCellsToRender`
- simulateScroll which triggers a call to `_updateCellsToRender` and calls setState
- setState callback is executed for the first `_updateCellsToRender` -> cells are not rendered, since at the time state.pendingScrollUpdateCount is 1
- setState callback is executed for the second `_updateCellsToRender` -> cells are rendered since state.pendingScrollUpdateCount is now 0

I will see how I can split this into multiple commits to make reviewing easier

janicduplessis commented:
@NickGerleman I've separated this in multiple commits, it should be easier to review now. Let me know if I forgot to address any of your points.

NickGerleman commented:
Have time blocked out to review this tomorrow 👍

NickGerleman commented:
Thanks for organizing this into commits. I think everything I had noticed has been addressed, and will import this.

janicduplessis commented:
@miafoo I noticed that the first issue you mentionned seems to happen only when there are few items in the list. I am not sure if it is related to the fix in this PR so I will look at it after.

NickGerleman commented:
> I am not sure if it is related to the fix in this PR so I will look at it after.

Is this something I should wait on testing before landing?

janicduplessis commented:
I don’t think so, it seems like this is a separate problem that is not related to this fix so I think we can go ahead and land this first and can have a follow up to fix that other issue.

janicduplessis commented:
Thanks @NickGerleman for your help with reviewing and getting this merged!
